### ISSUE TITLE
Compilation succeeds, executable segfaults: a probelm with *.ibc dependencies?

### ISSUE THREAD

**Issue by nicolabotta** [idris-lang/Idris-dev#2871]
From a fresh copy of https://github.com/nicolabotta/SeqDecProbs/tree/master/frameworks/14-, it should be possible to do the following:

1) Compile NonNegRationalTest.lidr with 

> idris +RTS -K32000000 -RTS -p contrib -p effects --warnreach NonNegRationalTest.lidr -o a.out

2) Start a.out. This should execute in less than one second and printout 

> 2067/616
> 32/11
> 3859/616

3) Edit NatGCDEuler.lidr and replace all occurrences of Sigma with Subset and all occurrences of MkSigma with Element

4) Type check NatGCDEuler.lidr and rebuild NatGCDEuler.ibc with

> idris +RTS -K32000000 -RTS -p contrib --warnreach NatGCDEuclid.lidr

5) Recompile NonNegRationalTest.lidr as above

Compilation in step 5 should succeed but the resulting executable segfault.

**Comment by melted** [comment]
This crashes for me with 0.9.20.2

**Comment by ahmadsalim** [comment]
@nicolabotta Could you kindly confirm whether this is still an issue? Thanks.

**Comment by nicolabotta** [comment]
@ahmadsalim Yes, I have just reproduced the results and can confirm that the issue persists under 0.11.2-git:4ed9c1b. 

If you try to reproduce the results, please notice that `NonNegRationalTest.lidr` has been renamed to `NonNegRationalBasicTests.lidr` and `NatGCDEuler.lidr` is in fact called `NatGCDEuclid.lidr`!

Thanks for looking into this issue! Please, also notice that the current Idris system does not seem to keep track of a modules's dependencies in the way I would expect it to do. For instance, I can compile `NonNegRationalBasicTests.lidr`:
```
idris +RTS -K32000000 -RTS -p contrib -p effects --warnreach NonNegRationalBasicTests.lidr -o a.out
```
, edit a file `NonNegRationalBasicTests.lidr` depends upon (for example `PNatOperations.lidr`) and recompile. This will fail with
```
Loading ./NonNegRationalBasicTests.ibc failed: Module needs reloading:
	SRC : "./PNatOperations.lidr"
	Modified at: 2016-06-04 16:06:18.924627286 UTC
	IBC : "./PNatOperations.ibc"
	Modified at: 2016-06-04 14:42:27.514848455 UTC
```
If I remember correctly, in this situations previous Idris systems would rebuild all the files `NonNegRationalBasicTests.lidr` depends upon that depend on `PNatOperations.lidr` and then rebuild `NonNegRationalBasicTests.lidr` itself. I might be wrong, of course.

**Comment by ahmadsalim** [comment]
@nicolabotta Great, thanks.
